packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -372,6 +378,8 @@ function collectTypeParameterUsageCounts(
 
 interface MappedType extends ts.ObjectType {
   typeParameter?: ts.Type;
+  constraintType?: ts.Type;
+  templateType?: ts.Type;
```

Comment:
Ah, thanks, this reminded me to file an issue on TypeScript: https://github.com/microsoft/TypeScript/issues/59221

Reply:
Did you see the comment on https://github.com/microsoft/TypeScript/issues/59221#issuecomment-2223721027? These properties are not safe to use if the checker hasn't actually touched the types to evaluate them for some reason.

Reply:
Is there a reliable way for a client to access the parts of a mapped type, then?

Reply:
I am not aware of any "official" way, but staring at the code, I _think_ one could call some other random checker method to get `resolveStructuredTypeMembers` to run on the mapped type, thus populating those fields (assuming the mapped types have `TypeFlags.Objejct`, anyway). So _maybe_ call `getPropertiesOfType` on it beforehand? We're in the danger zone with this...

Reply:
Assuming `getPropertiesOfType()` does the trick, then this code should work as intended.

Here's the relevant code:

https://github.com/typescript-eslint/typescript-eslint/blob/16392ab7a4503843ca3cc59831102971754ab9cc/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L385-L387

https://github.com/typescript-eslint/typescript-eslint/blob/16392ab7a4503843ca3cc59831102971754ab9cc/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L273-L284

Even our implementation of `isMappedType` relies on `typeParameter` being populated. But a few lines earlier we call `type.getProperties()`, which is a [thin wrapper][1] around `getPropertiesOfType`, so we've done what we can to make that the case.

[1]: https://github.com/microsoft/TypeScript/blob/edd08a570a755e00fd58b971aa4a639545ef3950/src/services/services.ts#L849-L851